### ISSUE TITLE
Discrepancies in refinement between p4est and Triangulation

### ISSUE THREAD

**Issue by marcfehling** [dealii/dealii#8647]
PR #8637 introduced a scenario in which the refinement and coarsening flags left after calling `prepare_coarsening_and_refinement()` do not represent how refinement will be performed by `p4est` during `execute_coarsening_and_refinement()`.

The fix introduced in this PR will ultimately rely on `p4est`'s refinement behavior. However as @bangerth stated, we are supposed to be predicting during `prepare_coarsening_and_refinement()` how `p4est` will perform refinement. This does not seem to be the case right now.

**Comment by marcfehling** [comment]
See also #8993 and #9036.

**Comment by tjhei** [comment]
Hey Marc, do you know what causes the difference between prepare and what p4est decides to do?

Something related has been on my todo list for a long time and I have (had?) a working implementation at some point. The idea was basically:
1. Turn off any smoothing/ 2-to-1 handling inside p4est.
2. Only use refinement flags on locally owned cells and perform "smoothing" (based on flags) correctly in parallel.
3. Exchange refinement flags in parallel and repeat 2&3 until we converge to a solution.

This might fix the issue at hand. If we hand a valid refinement request to p4est, p4est should be able to perform exactly that.

**Comment by marcfehling** [comment]
Thank you for pointing out your idea! I am still not sure how this happens, but I ran into this problem again just today: `p4est` refined cells that did not have a `refine_flag` set from the `deal.II` side.

For now, I will continue to tie critical operations to the `p::d::Triangulation` callbacks, since the oracle specifies which cells will be refined in the end. I need to postpone tackling this issue for now - finishing my thesis becomes acute...
